Fix `track_order` usage in do block

t-bltg opened the pull request:
Follow up of https://github.com/JuliaIO/HDF5.jl/pull/912.

- fix https://github.com/JuliaIO/HDF5.jl/issues/939;
- fix https://github.com/JuliaIO/HDF5.jl/issues/985.

Remove `IDX_TYPE` and `ORDER` globals (non thread safe global states).

~~And a little cleanup / simplifications~~.

t-bltg commented on test/runtests.jl:
```diff
@@ -12,51 +13,27 @@ Pkg.develop(PackageSpec(path=joinpath(filter_path, "H5Zzstd")))
-@testset "HDF5.jl" begin
-
-@debug "plain"
 include("plain.jl")
```
The same can be achieved with the removal of the `@testset "HDF5.jl"` block, so that specific `Test Summary: | Pass  Broken  Total` appear when using `include`. I believe that the top level `@testset` doesn't add anything and has the drawback of hiding the test sequence. With the removal you also track down hangs: https://github.com/JuliaIO/HDF5.jl/runs/7194199530?check_suite_focus=true#step:5:159.

mkitti replied:
Yes, I understand. However, we only really wanted it enabled during debugging and for CI. The multiprocess setup of `Pkg.test` also creates some unusual interference. Let's discuss in another pull request, please.

t-bltg commented on src/file.jl:
```diff
@@ -1,5 +1,5 @@
 """
-    h5open(filename::AbstractString, mode::AbstractString="r"; swmr=false, pv...)
+    h5open(filename::AbstractString, mode::AbstractString = "r"; swmr = false, pv...)
```
Already removed after https://github.com/JuliaIO/HDF5.jl/pull/983#issuecomment-1175011022.

mkitti commented on src/file.jl:
```diff
@@ -64,24 +64,37 @@ function h5open(filename::AbstractString, mode::AbstractString = "r";
     end
 end
 
+_change_iteration_order(track_order::Bool) = if track_order
```
This formatting is a bit weird. Just use the long form function syntax here so the indentation helps indicate where the function begins and ends.

I'm also not clear why it was necessary to move the definition. The order of method definitions should not matter.

t-bltg replied:
> Oh, I see that it was formatted like that before.

I'll change that to long def.

> Anyways, is this change actually necessary?

Yes, because `src/fileio.jl` is conditionally loaded by `Requires.jl`: https://github.com/JuliaIO/HDF5.jl/blob/ccdb3e90a5910bee954d066de7b3f775f9067bc3/src/HDF5.jl#L208.

mkitti reviewed (changes requested):
Did the method definitions need to move?

t-bltg commented:
> Did the method definitions need to move?

Yes, because `src/fileio.jl` is conditionally loaded by `Requires.jl`: https://github.com/JuliaIO/HDF5.jl/blob/ccdb3e90a5910bee954d066de7b3f775f9067bc3/src/HDF5.jl#L208.

mkitti commented on test/fileio.jl:
```diff
@@ -50,6 +50,12 @@ let fn = tempname() * ".h5"
+  h5open(fn, "r"; track_order=true) do io
+    # display(io)
+    @test all(keys(io) .== ["b", "a", "G"])
+  end
```
Test that the order actually reverts after the `do` block.

t-bltg replied:
Will do.

mkitti commented:
We may eventually want to move `IDX_TYPE` into `task_local_storage`, but that is outside the scope of this pull request.

t-bltg commented:
> We may eventually want to move IDX_TYPE into task_local_storage, but that is outside the scope of this pull request.

Not needed, I'm reworking the pr to remove unsafe globals `IDX_TYPE` and `ORDER`.

mkitti commented:
> Not needed, I'm reworking the pr to remove unsafe globals `IDX_TYPE` and `ORDER`.

Honestly, I think this is pretty close to mergeable. Perhaps this second refactor should be another pull request?

Here, I'm mainly waiting on someone else, perhaps @musm to review before merging.

mkitti reviewed (dismissed):
This is close to mergeable. The one thing that is missing is checking the reversion of `IDX_TYPE` after the `do` syntax.

mkitti commented:
`track_order` seems to conflate attribute order and link order. However, a dataset only has attribute order.

Thus you need to have a distinct implementation of the track_order accessors for dataset creation property lists.

This makes me wonder what happens when link track order and attribute track order differ.

mkitti commented on src/types.jl:
```diff
@@ -38,9 +38,10 @@ end
 
-    function File(id, filename, toclose::Bool=true)
-        f = new(id, filename)
+    function File(id, filename, fcpl::FileCreateProperties=FileCreateProperties(), toclose::Bool=true)
```
Since we are getting a file id as an argument, shouldn't we get the `FileCreateProperty` from the file id via `H5Fget_create_plist`?

https://docs.hdfgroup.org/hdf5/develop/group___h5_f.html#ga2f823a9e929b00b06a6be80619a61778

Also why do we need to add a `track_order` property to the File struct?  Should it basically be a property of the root group?

t-bltg replied:
See the comments, below this commit was obsolete, I've updated the PR.

t-bltg commented:
> Thus you need to have a distinct implementation of the track_order accessors for dataset creation property lists.

Yes, or support `track_order` only for `Group`s and `File`s.

I'm currently struggling with the property lists for `Group`s and `File`s **when reading** since `get_create_properties(...)` through `h5g_get_create_plist` or `h5f_get_create_plist` doesn't seem to return the correct values.
Currently, `fcpl` is used only upon writing a new file: https://github.com/JuliaIO/HDF5.jl/blob/ccdb3e90a5910bee954d066de7b3f775f9067bc3/src/file.jl#L35
thus the `track_order` property of `fcpl` is lost when reading.

It doesn't seem to be possible to modify the property list after the call to `h56f_open` here: https://github.com/JuliaIO/HDF5.jl/blob/ccdb3e90a5910bee954d066de7b3f775f9067bc3/src/file.jl#L43
This is why I've added a `fcpl` or `gcpl` named field to the `File` and `Group` structs and try to work with that now.

What remains is a failing test which occurs `h5object` is used for iterating the `File` objects:
https://github.com/JuliaIO/HDF5.jl/blob/ccdb3e90a5910bee954d066de7b3f775f9067bc3/src/objects.jl#L25

t-bltg commented:
The tests are now passing, although I'm not happy with the implementation which might be brittle.
An alternate view on this PR would help to see if we can make it bulletproof.

This would be breaking for `JLD.jl` (need to fix https://github.com/JuliaIO/JLD.jl/blob/14e66d98e1e31d1694ff7acf4139d5add72e4fa3/src/JLD.jl#L186).

mkitti commented:
If we broke the API for JLD, then we probably broke the API for someone else. This would not be mergeable until we decide to release v0.17 at least.

mkitti commented:
If we broke the API for JLD, then we probably broke the API for someone else. This would not be mergeable until we decide to release v0.17 at least.

mkitti commented:
My suggestion is to recover the commit where I said this was almost mergeable and submit that as another pull request. Merging this version is going to take some time.

t-bltg commented:
> If we broke the API for JLD, then we probably broke the API for someone else.

I can make this non-breakable by moving the positional argument after `toclose`, but that would be really odd since `toclose` is not a member of `File` and `fcpl` is ...

t-bltg commented:
> My suggestion is to recover the commit where I said this was almost mergeable and submit that as another pull request.

I honestly think that https://github.com/JuliaIO/HDF5.jl/pull/982/commits/1669dd3c708f64d44d256cbc14bd3597937e8648 is not viable either: I'd rather remove `IDX_TYPE` and `ORDER`.

> Merging this version is going to take some time.

I'd rather wait and polish than merge something wobbly.

mkitti reviewed (commented):
Thanks for attempting to address this.

The introduction of a mutable create property list for `File` and `Group` for use post-creation will require a further high-level design and planning review. The API modifications are breaking at multiple points and will thus need a minor release bump. My approval is withdrawn for now.

t-bltg commented:
> We'll have to discuss whether this removal alone is breaking.

Technically no, because `IDX_TYPE` and `ORDER` were never exported. I introduced them in https://github.com/JuliaIO/HDF5.jl/pull/912 in order to bypass fixed arguments of `h5..._iterate` as it was hard-coded previously, e.g.: https://github.com/JuliaIO/HDF5.jl/blob/6e8fc9242d12be7e2be5fb5407b9d272b9fedee9/src/show.jl#L211
This was an improvement at the time (over hard-coded arguments), but we might in the future need to change the indexing type and/or the order in a more robust and flexible manner.

> Additionally, I think the semantics of how track_order works is fundamentally changing here.

They need to, because https://github.com/JuliaIO/HDF5.jl/pull/912 was incomplete, as shown by https://github.com/JuliaIO/HDF5.jl/issues/939.

> There are other global variables that need to be addressed as well

Do you have an example ?

mkitti commented:
> Do you have an example ?

We basically need a structure that looks like this that contains property lists for everything. These will then serve as templates for the default property lists created by other functions. We can have a global one, and then perhaps provide a mechanism to provide a more local one.

```julia
struct HDF5Context
   attribute_access::HDF5.AttributeAccessProperties
   attribute_create::HDF5.AttributeCreateProperties
   dataset_access  ::HDF5.DatasetAccessProperties
   dataset_create  ::HDF5.DatasetCreateProperties
   dataset_transfer::HDF5.DatasetTransferProperties
   datatype_access ::HDF5.DatatypeAccessProperties
   datatype_create ::HDF5.DatatypeCreateProperties
   file_access     ::HDF5.FileAccessProperties
   file_create     ::HDF5.FileCreateProperties
   file_mount      ::HDF5.FileMountProperties
   group_access    ::HDF5.GroupAccessProperties
   group_create    ::HDF5.GroupCreateProperties
   link_access     ::HDF5.LinkAccessProperties
   link_create     ::HDF5.LinkCreateProperties
   object_copy     ::HDF5.ObjectCopyProperties
   object_create   ::HDF5.ObjectCreateProperties
   string_create   ::HDF5.StringCreateProperties
end

function HDF5Context()
   HDF5Context(
       HDF5.AttributeAccessProperties(),
       HDF5.AttributeCreateProperties(),
       HDF5.DatasetAccessProperties(),
       HDF5.DatasetCreateProperties(),
       HDF5.DatasetTransferProperties(),
       HDF5.DatatypeAccessProperties(),
       HDF5.DatatypeCreateProperties(),
       HDF5.FileAccessProperties(),
       HDF5.FileCreateProperties(),
       HDF5.FileMountProperties(),
       HDF5.GroupAccessProperties(),
       HDF5.GroupCreateProperties(),
       HDF5.LinkAccessProperties(),
       HDF5.LinkCreateProperties(),
       HDF5.ObjectCopyProperties(),
       HDF5.ObjectCreateProperties(),
       HDF5.StringCreateProperties()
   )
end
````

We could then use `task_local_storage` to store and retrieve this context. Functions would then use `H5Pcopy` to get their property lists. We could also just reference these default template property lists for `track_order`.

```julia
julia> context.file_create.track_order
false

julia> context.file_create.track_order = true
true

julia> task_local_storage(:hdf5_context, context) do
           println(task_local_storage(:hdf5_context).file_create.track_order)
       end
true
```

t-bltg commented:
> We basically need a structure that looks like this that contains property lists for everything. These will then serve as templates for the default property lists created by other functions. We can have a global one, and then perhaps provide a mechanism to provide a more local one.

Thanks for the example. It still feels a bit odd to use `task_local_storage`, but it seems to work and pass the tests.

mkitti reviewed (changes requested):
Thanks for implementing the context. Here's some feedback.

Ultimately, we'll need someone else to review a well.

t-bltg commented on src/HDF5.jl:
```diff
@@ -41,11 +41,10 @@ Filters, Drivers
 
-const IDX_TYPE = Ref(API.H5_INDEX_NAME)
-const ORDER = Ref(API.H5_ITER_INC)
+const CONTEXT = HDF5Context()
```
Moved to `properties.jl`.

t-bltg commented on src/properties.jl:
```diff
@@ -270,7 +270,7 @@ end
 get_track_order(p::Properties) = API.h5p_get_link_creation_order(p) != 0 && API.h5p_get_attr_creation_order(p) != 0
 
-function set_track_order(p::Properties, val::Bool)
+function set_track_order!(p::Properties, val::Bool)
```
Why ? These are unexported. I've renamed it to `set_track_order!` for consistency with e.g. `set_obj_track_times!`.

mkitti replied:
`export` does not define what is valid API or not. Non-exported functions could be public API if documented. That's why a package such as [PublicAPI.jl](https://docs.juliahub.com/PublicAPI/Jfs06/0.1.0/) exists.

While this is not documented, I would lean towards using `@deprecate` when possible. I agree this should change, but I also think we should deprecate should since `set_track_order` has existed across several releases.

t-bltg commented on src/HDF5.jl:
```diff
@@ -157,28 +156,28 @@ function Base.setindex!(parent::Union{File,Group}, val, path::Union{AbstractStri
+    property = getfield(local_context, name)
+    isvalid(property) ? property : nothing
+end
+get_access_properties(d::Dataset)   = something(get_global_property(:dataset_access), DatasetAccessProperties(API.h5d_get_access_plist(d)))
+get_access_properties(f::File)      = something(get_global_property(:file_access), FileAccessProperties(API.h5f_get_access_plist(f)))
+get_create_properties(d::Dataset)   = something(get_global_property(:dataset_create), DatasetCreateProperties(API.h5d_get_create_plist(d)))
+get_create_properties(g::Group)     = something(get_global_property(:group_create), GroupCreateProperties(API.h5g_get_create_plist(g)))
+get_create_properties(f::File)      = something(get_global_property(:file_create), FileCreateProperties(API.h5f_get_create_plist(f)))
+get_create_properties(a::Attribute) = something(get_global_property(:attribute_create), AttributeCreateProperties(API.h5a_get_create_plist(a)))
```
> I only want the global property if the object (Dataset, File, etc) is not valid.

Mmh no, this will make `track_order` fail since `API.h5f_get_create_plist(f)` returns a `fpcl` with incorrect `track_order`, although valid fpcl. A valid global context property should preempt object properties.

mkitti replied:
No. I would just copy that line into gen/api_defs.jl, run `julia --project=. gen/gen_wrappers.jl`, and then build a high level interface to copy property lists. I will probably add a bunch of API defs in another pull request if you just would like to wait for that.

mkitti replied:
> Mmh no, this will make `track_order` fail since `API.h5f_get_create_plist(f)` returns a `fpcl` with incorrect `track_order`, although valid opened `fpcl`. A valid global context property should preempt object properties.

Restore the old methods then and create a new family of methods. `get_access_from_context` or `get_context_create` or `get_context_access` perhaps?

t-bltg commented on src/properties.jl:
```diff
@@ -874,3 +874,45 @@ const UTF8_ATTRIBUTE_PROPERTIES = AttributeCreateProperties()
+   object_copy     ::HDF5.ObjectCopyProperties
+   object_create   ::HDF5.ObjectCreateProperties
+   string_create   ::HDF5.StringCreateProperties
+end
```
No, see https://github.com/JuliaIO/HDF5.jl/pull/982#discussion_r918227986.

t-bltg commented on src/file.jl:
```diff
@@ -56,32 +56,39 @@ function h5open(filename::AbstractString, mode::AbstractString = "r";
-        f(file)
-    finally
-        close(file)
+function h5open(f::Function, args...; pv...)
+    file = h5open(args...; pv...)
+    task_local_storage(:hdf5_context, CONTEXT) do
+        ctx = task_local_storage(:hdf5_context)
+        if (track_order = get(pv, :track_order, nothing)) !== nothing
+            ctx.file_create.track_order = ctx.group_create.track_order = track_order
+        end
+        try
+            f(file)
+        finally
+            close(file)
+            close(ctx.file_create)
+            close(ctx.group_create)
+        end
     end
 end
```
Ok fine, but I have no idea how to implement the `Properties` copy.

t-bltg commented on src/file.jl:
```diff
@@ -42,7 +42,7 @@ function h5open(filename::AbstractString, mode::AbstractString, fapl::FileAccess
         fid = API.h5f_open(filename, flag, fapl)
     end
-    return File(fid, filename)
+    File(fid, filename)
```
They are superfluous, but fine.

mkitti commented:
I resolved a few line conflicts in 9585bd3.

musm commented on src/properties.jl:
```diff
@@ -874,3 +874,52 @@ const UTF8_ATTRIBUTE_PROPERTIES = AttributeCreateProperties()
 _link_properties(::AbstractString) = UTF8_LINK_PROPERTIES
 _attr_properties(::AbstractString) = UTF8_ATTRIBUTE_PROPERTIES
+
+struct HDF5Context
```
HDF5 prefix below is unnecessary

musm commented on src/properties.jl:
```diff
@@ -874,3 +874,52 @@ const UTF8_ATTRIBUTE_PROPERTIES = AttributeCreateProperties()
+   )
+end
+
+function Base.copy(x::HDF5Context)
```
Is the `copy` method ever called in the current implementation, if so it seems like this wouldn't work without and PR  https://github.com/JuliaIO/HDF5.jl/pull/995 is needed for a correct implementation   (cc @mkitti )

t-bltg replied:
Rebased against master for https://github.com/JuliaIO/HDF5.jl/pull/995.

musm commented on src/file.jl:
```diff
@@ -56,32 +56,39 @@ function h5open(filename::AbstractString, mode::AbstractString = "r";
+function h5open(f::Function, args...; context = copy(CONTEXT), pv...)
+    file = h5open(args...; pv...)
+    task_local_storage(:hdf5_context, context) do
+        ctx = task_local_storage(:hdf5_context)
```
This is kinda confusing why do you need to use `ctx` instead of the passed in `context` ?

t-bltg replied:
If I understood `tls` correctly, this construction allow access of the context from withing a task (private). Anyway, this was a request of @mkitti to use tls so I will let him handle it.

t-bltg commented on src/types.jl:
```diff
@@ -138,3 +138,9 @@ const ScalarType = Union{BitsType,Reference}
 const Object = Union{Group,Dataset,Datatype}
+
+idx_type(obj::Union{File,Group}) = get_track_order(get_create_properties(obj)) ? API.H5_INDEX_CRT_ORDER : API.H5_INDEX_NAME
+idx_type(obj::Any) = API.H5_INDEX_NAME
```
No, not needed. I'm afk today so I will rework this pr tomorrow.

musm commented on test/fileio.jl:
```diff
@@ -36,20 +36,41 @@ end
   h5open(fn, "w"; track_order=true) do io
-    fcpl = HDF5.get_create_properties(io)
-    @test fcpl.track_order
+    @test HDF5.get_create_properties(io).track_order
```
is this a NFC (non functional change)?

t-bltg replied:
Rationale: consistency with the test written after. But I won't argue over a 2 lines change. Restored.

musm commented:
Generally looks good, still reviewing some of the PR. 

Could you give a brief summary of  how the task local changes are an improvement and operate?

One crude way to probe the value of the order would be to have global const refs, but I could easily imagine those values getting garbled with repeated calls in independent threads, even if things don't work 100% in multi-threaded contexts yet when using HDF5

mkitti commented:
The task local storage (TLS) is actually my idea, so I'll explain it. The (TLS) is basically a global namespace specific to each `Task`. If you start a new `Task`, you have a new `TLS`.

In our case, we need to temporarily change the global space for some fixed duration. To do so we can launch a new `Task` and then populate the TLS with the modified "global" context. Functions running within the `Task` will see the modified variables. These changes will be undone upon completion of the `Task`. 

Specifically in our case those local variables are property lists and their properties. In this PR, we are mainly thinking about index order. In the future, I think we should use these property lists as a template when creating new property lists. When we need a new property list, we copy it from the local context rather than creating one _de novo_.

This is better than using a global constant `Ref` to store state in that the changes are never seen outside the `Task`. Although multithreading does not currently apply here, this also happens to be compatible with that case.

I have a working draft here of what I think this should look like:

https://github.com/mkitti/HDF5.jl/blob/be385088adef7d1a83b994b0b140e75b530f7272/src/context.jl

mkitti commented on src/HDF5.jl:
```diff
@@ -157,27 +153,28 @@ function Base.setindex!(parent::Union{File,Group}, val, path::Union{AbstractStri
+    property = getfield(local_context, name)
+    isvalid(property) ? property : nothing
+end
+get_access_properties(d::Dataset)   = something(get_context_property(:dataset_access), DatasetAccessProperties(API.h5d_get_access_plist(d)))
+get_access_properties(f::File)      = something(get_context_property(:file_access), FileAccessProperties(API.h5f_get_access_plist(f)))
+get_create_properties(d::Dataset)   = something(get_context_property(:dataset_create), DatasetCreateProperties(API.h5d_get_create_plist(d)))
+get_create_properties(g::Group)     = something(get_context_property(:group_create), GroupCreateProperties(API.h5g_get_create_plist(g)))
+get_create_properties(f::File)      = something(get_context_property(:file_create), FileCreateProperties(API.h5f_get_create_plist(f)))
+get_create_properties(a::Attribute) = something(get_context_property(:attribute_create), AttributeCreateProperties(API.h5a_get_create_plist(a)))
```
```suggestion
get_access_properties(d::Dataset)   = DatasetAccessProperties(API.h5d_get_access_plist(d))
get_access_properties(f::File)      = FileAccessProperties(API.h5f_get_access_plist(f))
get_create_properties(d::Dataset)   = DatasetCreateProperties(API.h5d_get_create_plist(d))
get_create_properties(g::Group)     = GroupCreateProperties(API.h5g_get_create_plist(g))
get_create_properties(f::File)      = FileCreateProperties(API.h5f_get_create_plist(f))
get_create_properties(a::Attribute) = AttributeCreateProperties(API.h5a_get_create_plist(a))
```

Revert this. Just call `get_context_property` directly if we are interested in getting a property from the context. If we want to choose between object and context properties, that should be further downstream on a per-property basis.

mkitti replied:
This has not been resolved...

mkitti commented on src/file.jl:
```diff
@@ -56,32 +56,38 @@ function h5open(filename::AbstractString, mode::AbstractString = "r";
-        f(file)
-    finally
-        close(file)
+function h5open(f::Function, args...; context = copy(CONTEXT), pv...)
+    file = h5open(args...; pv...)
+    task_local_storage(:hdf5_context, context) do
+        ctx = task_local_storage(:hdf5_context)
+        if (track_order = get(pv, :track_order, nothing)) !== nothing
+            ctx.file_create.track_order = ctx.group_create.track_order = track_order
+        end
+        try
+            f(file)
+        finally
+            close(file)
+            close(ctx)
+        end
```
```suggestion
function h5open(f::Function, args...; context = copy(CONTEXT), pv...)
    file = h5open(args...; pv...)
    task_local_storage(:hdf5_context, context) do
        if (track_order = get(pv, :track_order, nothing)) !== nothing
            context.file_create.track_order = track_order
            context.group_create.track_order = track_order
        end
        try
            # The function body can access `context` via `get_context_property`
            f(file)
        finally
            close(file)
            close(context)
        end
```

mkitti commented on src/types.jl:
```diff
@@ -138,3 +138,9 @@ const ScalarType = Union{BitsType,Reference}
 # Define an H5O Object type
 const Object = Union{Group,Dataset,Datatype}
+
+idx_type(obj::Union{File,Group}) = get_track_order(get_create_properties(obj)) ? API.H5_INDEX_CRT_ORDER : API.H5_INDEX_NAME
```
```suggestion
idx_type(obj::File) = get_context_property(:file_create).track_order ? API.H5_INDEX_CRT_ORDER : API.H5_INDEX_NAME
idx_type(obj::Group) = get_context_property(:group_create).track_order ? API.H5_INDEX_CRT_ORDER : API.H5_INDEX_NAME
```

mkitti replied:
Perhaps we should always use `H5_INDEX_CRT_ORDER` if either the context or object property is enabled? Anyways, any fallback logic should be implemented here rather than `get_create_property` which has other uses.

```suggestion
idx_type(obj::File) =
    get_context_property(:file_create).track_order ||
    get_create_property(obj).track_order ? 
    API.H5_INDEX_CRT_ORDER : API.H5_INDEX_NAME
idx_type(obj::Group) =
    get_context_property(:group_create).track_order ||
    get_create_property(obj).track_order ? 
    API.H5_INDEX_CRT_ORDER : API.H5_INDEX_NAME
```

mkitti commented on src/objects.jl:
```diff
@@ -22,10 +22,16 @@ end
-    obj_type == API.H5I_DATATYPE ? Datatype(obj_id, file(parent)) :
-    obj_type == API.H5I_DATASET ? Dataset(obj_id, file(parent)) :
-    error("Invalid object type for path ", path)
+    obj = if obj_type == API.H5I_GROUP
+        Group(obj_id, file(parent))
+    elseif obj_type == API.H5I_DATATYPE
+        Datatype(obj_id, file(parent))
+    elseif obj_type == API.H5I_DATASET
+        Dataset(obj_id, file(parent))
+    else
+        error("Invalid object type for path ", path)
+    end
+    obj
```
```suggestion
    obj_type == API.H5I_GROUP ? Group(obj_id, file(parent)) :
    obj_type == API.H5I_DATATYPE ? Datatype(obj_id, file(parent)) :
    obj_type == API.H5I_DATASET ? Dataset(obj_id, file(parent)) :
    error("Invalid object type for path ", path)
```

This is an unrelated stylistic change. If we are going to do this, we should do this globally rather than just right here. properties.jl uses the existing style a lot.

t-bltg replied:
Probably leftovers from developing. Restored.

mkitti commented on src/HDF5.jl:
```diff
@@ -157,27 +153,28 @@ function Base.setindex!(parent::Union{File,Group}, val, path::Union{AbstractStri
+function get_context_property(name::Symbol)
+    local_context = get(task_local_storage(), :hdf5_context, CONTEXT)
+    property = getfield(local_context, name)
+    isvalid(property) ? property : nothing
```
Perhaps we could do this in the future in another pull request?

```
get_create_properties(::Type{File}) = get_context_property(:file_create)
```

mkitti replied:
`get_context_property` should move into `context.jl`.

t-bltg replied:
> After https://github.com/JuliaIO/HDF5.jl/pull/993 we do not need this. Property lists are now auto-initialized.

Did you try it ? Removing it fails the added tests.

> get_context_property should move into context.jl.

Moved.

mkitti replied:
I also changed the tests. If you would like me to add the commit, I would be happy to do so.

t-bltg replied:
Reworked.

mkitti commented on src/properties.jl:
```diff
@@ -875,3 +875,53 @@ const UTF8_ATTRIBUTE_PROPERTIES = AttributeCreateProperties()
 _link_properties(::AbstractString) = UTF8_LINK_PROPERTIES
 _attr_properties(::AbstractString) = UTF8_ATTRIBUTE_PROPERTIES
+
+struct HDF5Context
```
We should move the context stuff to it's own file, `context.jl`.

t-bltg replied:
Moved.

mkitti commented on test/fileio.jl:
```diff
@@ -50,6 +50,29 @@ let fn = tempname() * ".h5"
+
+  # issue #939
+  h5open(fn, "r"; track_order=true) do io
+    @test HDF5.get_create_properties(io).track_order
```
```suggestion
    @test HDF5.get_context_property(:file_create).track_order
```

mkitti replied:
This is clearer to me that the property is coming from the context rather than the object instance `io`.

mkitti commented on test/fileio.jl:
```diff
@@ -50,6 +50,29 @@ let fn = tempname() * ".h5"
+  h5open(fn, "r"; track_order=true) do io
+    @test HDF5.get_create_properties(io).track_order
+    @test all(keys(io) .== ["b", "a", "G"])
+    @test HDF5.get_create_properties(io["G"]).track_order
```
```suggestion
    @test HDF5.get_context_property(:group_create).track_order
```

mkitti commented on test/fileio.jl:
```diff
@@ -50,6 +50,29 @@ let fn = tempname() * ".h5"
+  end
+
+  h5open(fn, "r"; track_order=false) do io
+    @test !HDF5.get_create_properties(io).track_order
```
```suggestion
    @test !HDF5.get_context_property(:file_create).track_order
```

mkitti commented on test/fileio.jl:
```diff
@@ -50,6 +50,29 @@ let fn = tempname() * ".h5"
+  h5open(fn, "r"; track_order=false) do io
+    @test !HDF5.get_create_properties(io).track_order
+    @test all(keys(io) .== ["G", "a", "b"])
+    @test !HDF5.get_create_properties(io["G"]).track_order
```
```suggestion
    @test !HDF5.get_context_property(:group_create).track_order
```

mkitti commented on test/fileio.jl:
```diff
@@ -50,6 +50,29 @@ let fn = tempname() * ".h5"
+  end
+
+  h5open(fn, "r") do io
+    @test !HDF5.get_create_properties(io).track_order
```
```suggestion
    @test !HDF5.get_context_property(:file_create).track_order
```

mkitti commented on test/fileio.jl:
```diff
@@ -50,6 +50,29 @@ let fn = tempname() * ".h5"
+  h5open(fn, "r") do io
+    @test !HDF5.get_create_properties(io).track_order
+    @test all(keys(io) .== ["G", "a", "b"])
+    @test HDF5.get_create_properties(io["G"]).track_order  # inferred from file, created with `track_order=true`
```
```suggestion
    @test HDF5.get_context_property(:group_create).track_order  # inferred from file, created with `track_order=true`
```

mkitti reviewed (changes requested):
Merge master to get #993 if you have not already. #993 auto-initializes the property lists. Thus we do not need a fallback to the object's property list since the context will always have a valid property list on demand.

If someone wants to change the global index type or order, they can modify the global `CONTEXT`. Otherwise, they can use the `do` syntax for `h5open` to locally change the value.

t-bltg commented:
Looks like the failure on `MPI` built `hdf5` is spurious:

https://github.com/JuliaIO/HDF5.jl/runs/7544045615?check_suite_focus=true ✔
https://github.com/JuliaIO/HDF5.jl/runs/7544232871?check_suite_focus=true ✘

Even on failure, the outcome is a passing test.

mkitti commented:
```
  MethodError: no method matching h5o_get_info_by_name(::HDF5.File, ::String, ::Base.RefValue{HDF5.API.H5O_info1_t}, ::Int64)
```

That method error seems real though.

t-bltg commented:
Yes, but unrelated to this PR, since also present on master:
https://github.com/JuliaIO/HDF5.jl/runs/7429724952?check_suite_focus=true

mkitti commented:
I'm going to try to squash the test failure on this branch.

mkitti commented:
I fixed the tests here in [f5910a6](https://github.com/JuliaIO/HDF5.jl/pull/982/commits/f5910a6cf8c99b03737c4cdd47d9d39e7bec98ab) and in #1000

musm commented:
> The task local storage (TLS) is actually my idea, so I'll explain it. The (TLS) is basically a global namespace specific to each `Task`. If you start a new `Task`, you have a new `TLS`.
> 
> In our case, we need to temporarily change the global space for some fixed duration. To do so we can launch a new `Task` and then populate the TLS with the modified "global" context. Functions running within the `Task` will see the modified variables. These changes will be undone upon completion of the `Task`.
> 
> Specifically in our case those local variables are property lists and their properties. In this PR, we are mainly thinking about index order. In the future, I think we should use these property lists as a template when creating new property lists. When we need a new property list, we copy it from the local context rather than creating one _de novo_.
> 
> This is better than using a global constant `Ref` to store state in that the changes are never seen outside the `Task`. Although multithreading does not currently apply here, this also happens to be compatible with that case.
> 
> I have a working draft here of what I think this should look like:
> 
> https://github.com/mkitti/HDF5.jl/blob/be385088adef7d1a83b994b0b140e75b530f7272/src/context.jl

Do you plan on opening a PR after this is merged with  your working draft as I find  that those changes are important improvements?

musm commented:
@t-bltg  I'd recommend rebasing on master and cleaning up the commits. Then we can finally finalize merging in this PR. I'm not sure it would be best to squash all the commits or preserve the logical changes into several commits with more descriptive names. In either case a rebase would be helpful.

mkitti commented:
> Do you plan on opening a PR after this is merged with your working draft as I find that those changes are important improvements?

I would probably start another branch and send that in. I would also like to try implementing some of the applications. In particular, copying the context property lists to create default property lists for object creation.

For now, I've just documented this as an internal API that's under development.

mkitti commented:
I would just squash it.

t-bltg commented:
> I fixed the tests here in [f5910a6](https://github.com/JuliaIO/HDF5.jl/pull/982/commits/f5910a6cf8c99b03737c4cdd47d9d39e7bec98ab) and in https://github.com/JuliaIO/HDF5.jl/pull/1000

Thanks for fixing.

> @t-bltg I'd recommend rebasing on master and cleaning up the commits.

Rebased.

musm commented on src/types.jl:
```diff
@@ -138,3 +138,16 @@ const ScalarType = Union{BitsType,Reference}
+idx_type(obj) = API.H5_INDEX_NAME
+
+# TODO: implement alternative iteration order ?
+order(obj) = API.H5_ITER_INC
```
I'd think a more descriptive name would be better. We might want to use `order` in the future for something else.

t-bltg replied:
Yes, left for another PR ;)

musm commented on test/fileio.jl:
```diff
@@ -50,6 +50,30 @@ let fn = tempname() * ".h5"
+  end
+
+  h5open(fn, "r") do io
+    @test !HDF5.get_create_properties(io).track_order
```
If track order is false can't the check on line 73 fail since the keys are unordered? Wouldn't an equality check using `Set` be safer?

t-bltg replied:
In this case, `track_order` from `file_create` is inferred from file so this test is correct.

musm commented:
Thanks for sticking to this and sorry for the long review!

t-bltg commented:
Thanks all for helping finalizing this PR.
